src/util/file_bytes_stream.rs

```diff
@@ -53,3 +68,168 @@ impl FileBytesStream {
         Body::wrap_stream(self)
     }
 }
+
+#[derive(PartialEq, Eq)]
+enum FileSeekState {
+    NeedSeek,
+    Seeking,
+    Reading,
+}
+
+/// Wraps a `tokio::fs::File`, and implements a stream of `Bytes`s reading a portion of the
+/// file given by `range`.
+pub struct FileBytesStreamRange {
+    file_stream: FileBytesStream,
+    seek_state: FileSeekState,
+    start_offset: u64,
+}
+
+impl FileBytesStreamRange {
+    /// Create a new stream from the given file and range
+    pub fn new(file: File, range: HttpRange) -> FileBytesStreamRange {
+        FileBytesStreamRange {
+            file_stream: FileBytesStream::new_with_limit(file, range.length),
+            seek_state: FileSeekState::NeedSeek,
+            start_offset: range.start,
+        }
+    }
+
+    fn without_initial_range(file: File) -> FileBytesStreamRange {
+        FileBytesStreamRange {
+            file_stream: FileBytesStream::new_with_limit(file, 0),
+            seek_state: FileSeekState::NeedSeek,
+            start_offset: 0,
+        }
+    }
+}
+
+impl Stream for FileBytesStreamRange {
+    type Item = Result<Bytes, IoError>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<Option<Self::Item>> {
+        let Self {
+            ref mut file_stream,
+            ref mut seek_state,
+            start_offset,
+        } = *self;
+        if *seek_state == FileSeekState::NeedSeek {
+            *seek_state = FileSeekState::Seeking;
+            if let Err(e) = Pin::new(&mut file_stream.file).start_seek(SeekFrom::Start(start_offset)) {
+                return Poll::Ready(Some(Err(e)));
+            }
+        }
+        if *seek_state == FileSeekState::Seeking {
+            match Pin::new(&mut file_stream.file).poll_complete(cx) {
+                Poll::Ready(Ok(..)) => *seek_state = FileSeekState::Reading,
+                Poll::Ready(Err(e)) => return Poll::Ready(Some(Err(e))),
+                Poll::Pending => return Poll::Pending,
+            }
+        }
+        Pin::new(file_stream).poll_next(cx)
+    }
+}
+
+
+impl FileBytesStreamRange {
+    /// Create a Hyper `Body` from this stream.
+    pub fn into_body(self) -> Body {
+        Body::wrap_stream(self)
+    }
+}
+
+/// Wraps a `tokio::fs::File`, and implements a stream of `Bytes`s reading multiple portions of
+/// the file given by `ranges` using a chunked multipart/byteranges response.  A boundary is
+/// required to separate the chunked components.
```

Comment:
Explain more about boundaries, the user should be aware that they need sufficient entropy and probably should be only ASCII.